tests/DCPS/RtpsRelay/STUN/StunClient.cpp

```diff
@@ -0,0 +1,289 @@
+  request.block = block.get();
+  OpenDDS::DCPS::Serializer serializer(block.get(), true);
+  if (!(serializer << request)) {
+    std::cerr << "ERROR: Failed to serialize request" << std::endl;
+    status = EXIT_FAILURE;
+    return false;
+  }
+
+  const ssize_t bytes_sent = socket.send(block->rd_ptr(), block->length(), remote);
+  if (bytes_sent != static_cast<ssize_t>(block->length())) {
+    std::cerr << "ERROR: Failed to send: " << strerror(errno) << std::endl;
+    status = EXIT_FAILURE;
+    return false;
+  }
+
+  return true;
+}
+
+bool recv(int& status,
+          ACE_SOCK_Dgram& socket,
+          OpenDDS::STUN::Message& response)
+{
+  iovec iov;
+  ACE_INET_Addr addr;
+  const ssize_t bytes_received = socket.recv(&iov, addr);
+  if (bytes_received <= 0) {
+    std::cerr << "ERROR: Failed to recv: " << strerror(errno) << std::endl;
+    status = EXIT_FAILURE;
+    return false;
+  }
+
+  const auto data_block =
+    new (ACE_Allocator::instance()->malloc(sizeof(ACE_Data_Block))) ACE_Data_Block(bytes_received, ACE_Message_Block::MB_DATA,
+                                                                                   static_cast<const char*>(iov.iov_base), 0, 0, 0, 0);
+  OpenDDS::DCPS::Message_Block_Shared_Ptr buffer(new ACE_Message_Block(data_block));
+  buffer->length(bytes_received);
+
+  response.block = buffer.get();
+  OpenDDS::DCPS::Serializer deserializer(buffer.get(), true);
+  if (!(deserializer >> response)) {
+    std::cerr << "ERROR: Failed to deserialize response" << std::endl;
+    status = EXIT_FAILURE;
+    return false;
+  }
+
+  return true;
+}
+
+bool test_success(int& status,
+                  ACE_SOCK_Dgram& socket,
+                  const ACE_INET_Addr& remote)
+{
+  std::cerr << __func__ << std::endl;
+  bool retval = true;
+  OpenDDS::STUN::Message request;
+  request.class_ = OpenDDS::STUN::REQUEST;
+  request.method = OpenDDS::STUN::BINDING;
+  generate_transaction_id(request);
+  request.append_attribute(OpenDDS::STUN::make_fingerprint());
+
+  if (!send(status, socket, remote, request)) {
+    return false;
+  }
+
+  OpenDDS::STUN::Message response;
+  if (!recv(status, socket, response)) {
+    return false;
+  }
+
+  if (response.class_ != OpenDDS::STUN::SUCCESS_RESPONSE) {
+    std::cerr << "ERROR: Incorrect message class" << std::endl;
+    status = EXIT_FAILURE;
+    retval = false;
+  }
+
+  if (response.method != request.method) {
+    std::cerr << "ERROR: Incorrect method" << std::endl;
+    status = EXIT_FAILURE;
+    retval = false;
+  }
+
+  if (response.transaction_id != request.transaction_id) {
+    std::cerr << "ERROR: Incorrect transaction id" << std::endl;
+    status = EXIT_FAILURE;
+    retval = false;
+  }
+
+  ACE_INET_Addr a;
+  if (!response.get_mapped_address(a)) {
+    std::cerr << "ERROR: no mapped address" << std::endl;
+    status = EXIT_FAILURE;
+    retval = false;
+  }
+
+  char buffer[256];
+  a.addr_to_string(buffer, 256);
```

Comment:
ACE uses TCHAR here.

Reply:
This wasn't resolved by the later commit because it mixes ACE_TCHAR with std::ostream.